### PR TITLE
Persist userlist and PGB config on controller restart 

### PR DESCRIPTION
**Pull request by WRFitch**
## Proposal
When we restart the k8s controller on which a PGB operator runs, we lose our containers and therefore our config. This PR checks the peer databag for config files if they aren't available in the filesystem. 

## Context
- We can't keep logs because they aren't stored anywhere, and if they were, they should probably live in COS. 

## Release Notes
- Check peer databag for config files if they aren't available in filesystem. 

## Testing
- Added an integration test that restarts the juju controller pod.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -59,13 +59,23 @@ def _on_start(self, event: StartEvent) -> None:
             event.defer()
             return
 
-        if (default_config := self.peers.get_cfg()) is None:
+        try:
+            # Try and get pgb config. If it already exists in the filesystem it'll be refreshed,
+            # and if it only exists in the peer databag, pull it and write it to filesystem.
+            config = self.read_pgb_config()
```
[the remark this answers is not in the document]
Comment (WRFitch) [reply src/charm.py]: Agreed, I've updated this comment to make more sense

File: src/charm.py [hunk 2]
```diff
@@ -269,11 +279,18 @@ def read_pgb_config(self) -> PgbConfig:
             PgbConfig object containing pgbouncer config.
```
[the remark this answers is not in the document]
Comment (WRFitch) [reply src/charm.py:282]: I've kept `read_pgb_config` and `peers.get_cfg` separate, so this is no longer necessary

File: src/relations/peers.py [hunk 3]
```diff
@@ -289,6 +289,13 @@ def update_auth_file(self, auth_file: str) -> None:
         self.set_secret("app", AUTH_FILE_DATABAG_KEY, auth_file)
         logger.debug("updated auth file in peer databag")
 
+    def get_auth_file(self) -> str:
```
Comment (marceloneppel) [inline src/relations/peers.py:292]: Is this code being used in any part of the other files? I couldn't find its usage.

**Review by marceloneppel — COMMENTED** [review]
Does killing the controller pod (to force it to be recreated) break the test run?

File: tox.ini [hunk 4]
```diff
@@ -68,7 +68,7 @@ commands =
 description = Run all integration tests
 deps =
     pytest
-    juju
+    juju==2.9.11 # juju 3.0.0 has issues with retrieving action results
```
Comment (taurus-forever) [inline tox.ini:71]: JFMI, why do we lock such an old juju version? 2.9.37 is the latest as for today.

**Review by taurus-forever — APPROVED** [review]
I have single question as JFMI only. LGTM. Tnx!

File: tox.ini (hunk 4 above)
Comment (WRFitch) [reply tox.ini:71]: This is the last 2.9 version of _libjuju_, which doesn't follow the same version as juju :)

**Review by marceloneppel — APPROVED** [review]

**Review by marceloneppel — DISMISSED** [review]

**Review by paulomach — APPROVED** [review]
Looks sane and remind me that we need to take a look at cases like this on mysqland

**Review by marceloneppel — APPROVED** [review]

**Pull request merged**